auth/store/mongo/device_tokens_repository.go

```diff
@@ -0,0 +1,53 @@
+package mongo
+
+import (
+	"context"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/errors"
+	structuredmongo "github.com/tidepool-org/platform/store/structured/mongo"
+)
+
+// deviceTokenRepo implements devicetokens.Repository, writing data to a
+// MongoDB collection.
+type deviceTokenRepo structuredmongo.Repository
+
+// Upsert will create or update the given Config.
+func (r *deviceTokenRepo) Upsert(ctx context.Context, doc *devicetokens.Document) error {
+	// The presence of UserID and TokenID should be enforced with a mongodb
+	// index, but better safe than sorry.
+	if doc.UserID == "" {
+		return errors.New("UserID may not be empty")
+	}
+	if doc.TokenKey == "" {
+		return errors.New("TokenID may not be empty")
+	}
+
+	opts := options.Update().SetUpsert(true)
+	f := bson.M{"tokenKey": doc.TokenKey, "userId": doc.UserID}
+	_, err := r.UpdateOne(ctx, f, bson.M{"$set": doc}, opts)
+	if err != nil {
+		return errors.Wrap(err, "upserting device token")
```

Comment:
Do you often grep your logs for "unable"?

Reply:
Is "failure to upsert" allowed? What about "failure during upsert"? "failed to upsert"? "problem upserting"? 

I guess I'm being flippant now, sorry... But is this really something useful? If it is, I'd love to have it documented, because that's the only way I can see myself having a chance to remember which specific terms are acceptable (feels a bit like a magic incantation). I'm being flippant again, sorry, but this feels very very arbitrary to me.

Reply:
Yes, lack of any formal documentation that describes the platform standards and conventions is on me (or, well, on my past self, at any rate). At the time it was just Jamie and me and there were many other priorities than writing documentation for potential future engineers. So, at this point, for better or worse, the "documentation" is the code.

In any case, much of any convention is purely arbitrary. There are lots of way to write "technically correct" code, but adding convention on top of that improves readability and understandability. 

Some of the existing platform conventions are fine and either correlates with or at least does not contradict Golang. However, some platform conventions do contradict the _current_ Golang conventions (but perhaps did not when platform was first developed). 

If there is a contradiction, I'd default to sticking with the platform convention because there may be a good reason for that convention. In this case, not really, other than it is engrained in my head "unable to <whatever you were trying to do>" is standard.  In other cases (i.e. explicitly not creating global error variables) has a very good reason behind not doing that.

So, if you don't want to change this to match existing conventions, that's fine. I disagree, but I won't hold up this PR.

Reply:
I've already made the change.